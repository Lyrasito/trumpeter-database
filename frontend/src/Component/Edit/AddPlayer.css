.newPlayer-container {
  padding-bottom: 80px;
  margin-top: 80px;
}

.newPlayer {
  width: 1200px;
  display: flex;
  justify-content: space-between;
  margin-left: 80px;
}

.player-input {
  box-sizing: border-box;
  height: 63px;
  width: 200px;
  border: 3px solid #666666;
  border-radius: 30px;
  background-color: #ffffff;
  display: block;
  font-size: 24px;
}

.submit {
  height: 60px;
  width: 162px;
  border-radius: 30px;
  background-color: #ffc600;
  cursor: pointer;
  color: #000000;
  font-family: "WorkSans";
  font-size: 24px;
  font-weight: 600;
  letter-spacing: 0;
  line-height: 35px;
  margin-top: 30px;
}
.message {
  padding-left: 80px;
}
@media only screen and (min-width: 1201px) {
  .newPlayer {
    width: 85%;
  }
}
