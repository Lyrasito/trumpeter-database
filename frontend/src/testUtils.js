const fakePlayer = () => ({
  id: 1,
  name: "Fake Player",
  city: "Fake City",
  startYear: 1234,
  endYear: 5678,
});

const fakePlayers = () => [
  {
    id: 1,
    name: "Fake Player",
    city: "Fake City",
    startYear: 1234,
    endYear: 5678,
  },
  {
    id: 2,
    name: "Fake Player 2",
    city: "Fake City 2",
    startYear: 2345,
    endYear: 6789,
  },
];

const fakeAlbums = () => [
  {
    id: 1,
    title: "Fake Album 1",
    year: 1234,
    genre: "fakeGenre1",
    player_id: 1,
  },
  {
    id: 2,
    title: "Fake Album 2",
    year: 1234,
    genre: "fakeGenre2",
    player_id: 1,
  },
];

const fakeAlbum = () => ({
  id: 1,
  title: "Fake Album 1",
  year: 1234,
  genre: "fakeGenre1",
  player_id: 1,
});

const fakeGenres = () => ["fakeGenre1", "fakeGenre2"];

export { fakeAlbum, fakeAlbums, fakePlayer, fakeGenres, fakePlayers };
