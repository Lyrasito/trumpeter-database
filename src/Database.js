const fetch = require("node-fetch");

const baseUrl = "http://localhost:4000/api";

const Database = {
  async getPlayers() {
    const url = `${baseUrl}/players`;
    const response = await fetch(url);
    const jsonResponse = await response.json();
    const players = await jsonResponse.players.map((player) => ({
      id: player.id,
      name: player.name,
      city: player.city,
      startYear: player.start_year,
      endYear: player.end_year,
      image: player.image,
    }));
    return players;
  },

  async searchPlayers(city, year, genre) {
    let url = "";
    if (city && !year && !genre) {
      url = `${baseUrl}/players/search?city=${city}`;
    } else if (year && !city && !genre) {
      url = `${baseUrl}/players/search?year=${year}`;
    } else if (genre && !city && !year) {
      url = `${baseUrl}/players/search?genre=${genre}`;
    } else if (city && year && !genre) {
      url = `${baseUrl}/players/search?city=${city}&year=${year}`;
    } else if (city && genre && !year) {
      url = `${baseUrl}/players/search?city=${city}&genre=${genre}`;
    } else if (genre && year && !city) {
      url = `${baseUrl}/players/search?genre=${genre}&year=${year}`;
    } else if (city && year && genre) {
      url = `${baseUrl}/players/search?city=${city}&year=${year}&genre=${genre}`;
    }
    const response = await fetch(url);
    const jsonResponse = await response.json();
    const players = await jsonResponse.players.map((player) => ({
      id: player.id,
      name: player.name,
      city: player.city,
      startYear: player.start_year,
      endYear: player.end_year,
      image: player.image,
    }));
    return players;
  },

  async searchByName(name) {
    const url = `${baseUrl}/players/search?name=${name}`;
    const response = await fetch(url);
    const jsonResponse = await response.json();
    const players = await jsonResponse.players.map((player) => ({
      id: player.id,
      name: player.name,
      city: player.city,
      startYear: player.start_year,
      endYear: player.end_year,
      image: player.image,
    }));
    return players;
  },

  async getPlayerAlbums(player) {
    const url = `${baseUrl}/players/${player.id}/albums`;
    const response = await fetch(url);
    const jsonResponse = await response.json();
    const albums = await jsonResponse.albums.map((album) => ({
      id: album.id,
      title: album.title,
      year: album.year,
    }));
    return albums;
  },

  async getPlayerGenres(player) {
    const url = `${baseUrl}/players/${player.id}/albums/genres`;
    const response = await fetch(url);
    const jsonResponse = await response.json();
    return jsonResponse.genre;
  },

  async addPlayer(name, city, startYear, endYear) {
    const url = `${baseUrl}/players`;
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        player: {
          name: name,
          city: city,
          startYear: startYear,
          endYear: endYear,
        },
      }),
    });

    const jsonResponse = await response.json();
    return jsonResponse.player;
  },

  async addAlbum(playerId, title, year, genre) {
    const url = `${baseUrl}/players/${playerId}/albums`;
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        album: {
          title: title,
          year: year,
          genre: genre,
          playerId: playerId,
        },
      }),
    });
    const jsonResponse = await response.json();
    console.log(jsonResponse);
  },
};

module.exports = Database;

/*  */
