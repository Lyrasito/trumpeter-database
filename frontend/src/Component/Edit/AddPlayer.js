import React from "react";
import Database from "../../Database";
import "./AddPlayer.css";

class AddPlayer extends React.Component {
  state = {
    name: "",
    city: "",
    startYear: "",
    endYear: "",
    newPlayer: {},
    error: null,
  };

  handleChange = (event) => {
    const { name, type, value } = event.target;
    const val = type === "number" ? parseFloat(value) : value;
    this.setState({ [name]: val });
  };

  addPlayer = async () => {
    try {
      const newPlayer = await Database.addPlayer(
        this.state.name,
        this.state.city,
        this.state.startYear,
        this.state.endYear
      );
      this.setState({
        newPlayer: newPlayer,
        name: "",
        city: "",
        startYear: "",
        endYear: "",
      });
    } catch (err) {
      this.setState({ error: err.message, newPlayer: {} });
    }
  };
  renderPlayer = () => {
    if (this.state.newPlayer.name) {
      return (
        <h3 className="message" data-test="success-message">
          You have added {this.state.newPlayer.name} to the database!
        </h3>
      );
    }
  };
  renderError = () => {
    if (this.state.error) {
      return (
        <h4 className="message" data-test="error-message">
          {this.state.error}
        </h4>
      );
    }
  };
  render() {
    return (
      <div className="newPlayer-container">
        <h3 className="message">Add a player to the database!</h3>
        <div className="newPlayer">
          <label>
            Name
            <input
              className="player-input"
              name="name"
              data-test="name"
              onChange={this.handleChange}
              value={this.state.name}
            ></input>
          </label>
          <label>
            City/Cities
            <input
              className="player-input"
              name="city"
              data-test="city"
              onChange={this.handleChange}
              value={this.state.city}
            ></input>
          </label>
          <label>
            Year career started
            <input
              className="player-input"
              name="startYear"
              data-test="startYear"
              type="number"
              onChange={this.handleChange}
              value={this.state.startYear}
            ></input>
          </label>
          <label>
            Year career ended
            <input
              className="player-input"
              name="endYear"
              data-test="endYear"
              type="number"
              onChange={this.handleChange}
              value={this.state.endYear}
            ></input>
          </label>

          <button type="submit" className="submit" onClick={this.addPlayer}>
            Add Player
          </button>
        </div>
        {this.renderPlayer()}
        {this.renderError()}
      </div>
    );
  }
}

export default AddPlayer;
