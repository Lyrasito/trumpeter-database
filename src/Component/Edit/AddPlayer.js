import React from "react";
import Database from "../../Database";
import "./AddPlayer.css";

class AddPlayer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: "",
      city: "",
      startYear: null,
      endYear: null,
      newPlayer: {},
      error: null,
    };
    this.getName = this.getName.bind(this);
    this.getCity = this.getCity.bind(this);
    this.getStartYear = this.getStartYear.bind(this);
    this.getEndYear = this.getEndYear.bind(this);
    this.addPlayer = this.addPlayer.bind(this);
    this.renderPlayer = this.renderPlayer.bind(this);
    this.renderError = this.renderError.bind(this);
  }

  getName(event) {
    this.setState({ name: event.target.value });
  }
  getCity(event) {
    this.setState({ city: event.target.value });
  }
  getStartYear(event) {
    this.setState({ startYear: event.target.value });
  }
  getEndYear(event) {
    this.setState({ endYear: event.target.value });
  }
  async addPlayer() {
    try {
      const newPlayer = await Database.addPlayer(
        this.state.name,
        this.state.city,
        this.state.startYear,
        this.state.endYear
      );
      this.setState({
        newPlayer: newPlayer,
        name: "",
        city: "",
        startYear: "",
        endYear: "",
      });
    } catch (err) {
      this.setState({ error: err.message, newPlayer: {} });
    }
  }
  renderPlayer() {
    if (this.state.newPlayer.name) {
      return (
        <h3 className="message">
          You have added {this.state.newPlayer.name} to the database!
        </h3>
      );
    }
  }
  renderError() {
    if (this.state.error) {
      return <h4 className="message">{this.state.error}</h4>;
    }
  }
  render() {
    return (
      <div className="newPlayer-container">
        <h3 className="message">Add a player to the database!</h3>
        <div className="newPlayer">
          <label>
            Name
            <input
              className="player-input"
              id="createName"
              onChange={this.getName}
              value={this.state.name}
            ></input>
          </label>
          <label>
            City/Cities
            <input
              className="player-input"
              id="createCity"
              onChange={this.getCity}
              value={this.state.city}
            ></input>
          </label>
          <label>
            Year career started
            <input
              className="player-input"
              id="createStartYear"
              onChange={this.getStartYear}
              value={this.state.startYear}
            ></input>
          </label>
          <label>
            Year career ended
            <input
              className="player-input"
              id="createEndYear"
              onChange={this.getEndYear}
              value={this.state.endYear}
            ></input>
          </label>

          <button type="submit" className="submit" onClick={this.addPlayer}>
            Add Player
          </button>
        </div>
        {this.renderPlayer()}
        {this.renderError()}
      </div>
    );
  }
}

export default AddPlayer;
