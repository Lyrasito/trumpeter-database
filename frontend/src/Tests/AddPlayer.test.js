import AddPlayer from "../Component/Edit/AddPlayer";
import React from "react";
import toJSON from "enzyme-to-json";
import { mount, shallow } from "enzyme";
import wait from "waait";
import { fakePlayer } from "../testUtils";
import Database from "../Database";

const player = fakePlayer();
Database.addPlayer = jest
  .fn()
  .mockResolvedValueOnce(player)
  .mockImplementation(() => {
    throw new Error("Please fill out all fields");
  });

describe("<AddPlayer />", () => {
  it("renders properly", () => {
    const wrapper = shallow(<AddPlayer />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
  it("adds a player to the database", async () => {
    const wrapper = mount(<AddPlayer />);
    const name = wrapper.find("[data-test='name']");
    const city = wrapper.find("[data-test='city']");
    const startYear = wrapper.find("[data-test='startYear']");
    const endYear = wrapper.find("[data-test='endYear']");
    const button = wrapper.find(".submit");

    name.instance().value = player.name;
    name.simulate("change");
    city.instance().value = player.city;
    city.simulate("change");
    startYear.instance().value = player.startYear;
    startYear.simulate("change");
    endYear.instance().value = player.endYear;
    endYear.simulate("change");

    button.simulate("click");
    await wait();
    wrapper.update();

    expect(wrapper.find("[data-test='success-message']").text()).toBe(
      `You have added ${player.name} to the database!`
    );

    expect(Database.addPlayer).toHaveBeenCalledWith(
      player.name,
      player.city,
      player.startYear,
      player.endYear
    );
  });
  it("renders an error if called with insufficient input", async () => {
    const wrapper = mount(<AddPlayer />);
    const name = wrapper.find("[data-test='name']");
    const button = wrapper.find(".submit");
    name.instance().value = player.name;
    name.simulate("change");
    button.simulate("click");
    await wait();
    wrapper.update();
    expect(wrapper.find("[data-test='error-message']").text()).toBe(
      "Please fill out all fields"
    );
  });
});
