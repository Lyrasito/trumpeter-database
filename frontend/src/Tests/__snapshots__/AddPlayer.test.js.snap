// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddPlayer /> renders properly 1`] = `
<div
  className="newPlayer-container"
>
  <h3
    className="message"
  >
    Add a player to the database!
  </h3>
  <div
    className="newPlayer"
  >
    <label>
      Name
      <input
        className="player-input"
        data-test="name"
        name="name"
        onChange={[Function]}
        value=""
      />
    </label>
    <label>
      City/Cities
      <input
        className="player-input"
        data-test="city"
        name="city"
        onChange={[Function]}
        value=""
      />
    </label>
    <label>
      Year career started
      <input
        className="player-input"
        data-test="startYear"
        name="startYear"
        onChange={[Function]}
        type="number"
        value=""
      />
    </label>
    <label>
      Year career ended
      <input
        className="player-input"
        data-test="endYear"
        name="endYear"
        onChange={[Function]}
        type="number"
        value=""
      />
    </label>
    <button
      className="submit"
      onClick={[Function]}
      type="submit"
    >
      Add Player
    </button>
  </div>
</div>
`;
